centrallix-sysdoc/OSDriver_Authoring.md

```diff
@@ -115,497 +170,1032 @@ License:  Copyright (C) 2001-2011 LightSys Technology Services. See LICENSE.txt
-
-Within the initialization function, the driver should initialize all necessary global variables and register itself with the OSML.  Global variables should all be placed inside a single global 'struct', which is normally named similarly to the driver's prefix, except normally in all uppercase.  Under no circumstances should global variables be accessed outside of the module, except via the module's functions.
+Using the example above, we can query from the database using a statement like `select :title from /OMSS_DB/JNetHelp/rows`, which will open a sybase driver instance, then open a query and repeatedly fetch rows, getting the `title` attribute from each row.
 
-To register with the OSML, the driver must first allocate an ObjDriver structure and fill in its contents.  
 
-    pObjDriver drv;
-
-    drv = (pObjDriver)nmMalloc(sizeof(ObjDriver));
+## II Interface
+This section describes the standard interface between the OSML and the ObjectSystem driver itself.  Every driver should implement certain required functions.  (**Note**: Many drivers "implement" some required functions to simply fail with a not implemented or not supported error.  For example, most database drivers "implement" `Read()` and `Write()` this way because database content should be queried, not read).  Various optional functions are also available, which a driver is not required to implement.
+
+<!-- TODO: Greg
+  --- Double check the information in this table. Some of it is missing,
+  --- and I had to guess a lot of it by looking at how various drivers were
+  --- implemented. Also, which functions are optional and which are required
+  --- seems very chaotic and I vaguely remember there being a driver that
+  --- does not implement a ton of "required" functions, so please double
+  --- check that this information is 100% correct.
+  --->
+The driver should implement an `Initialize()` function, as well as the following (* indicates required functions):
+| Function Name                                             | Description
+| --------------------------------------------------------- | ------------
+| [Open](#function-open)*                                   | Opens a new driver instance object on a given root node.
+| [OpenChild](#function-openchild)                          | ???
+| [Close](#function-close)*                                 | Close an open object created by either `Open()` or `QueryFetch()`.
+| [Create](#function-create)                                | Create a new driver root node object.
```

Comment:
This is actually currently unused, as the OSML calls the driver Open instead with the `O_WRONLY | O_CREAT | O_EXCL` options to handle an objCreate call.

Reply:
This information has been added.